ribbon/src/main/java/feign/ribbon/RibbonClient.java

```diff
@@ -57,12 +64,33 @@ public Response execute(Request request, Request.Options options) throws IOExcep
     }
   }
 
-  private LBClient lbClient(String clientName) {
-    IClientConfig config = ClientFactory.getNamedConfig(clientName);
-    ILoadBalancer lb = ClientFactory.getNamedLoadBalancer(clientName);
-    return new LBClient(delegate, lb, config);
+  /**
+   * Returns a cached version of the {@link LBClient} if it exists otherwise a new {@link LBClient}
+   * is created and added to the cache.
+   * <p>
+   * Ribbon does not expect multiple LBClient instances for the same client. This introduces a cache
+   * to avoid redundantly creating them.
+   * 
+   * @param clientName - The clientName that the {@link LBClient} will be cache against
+   * @return
+   */
+  private LBClient getOrCreateLBClient(String clientName) {
+    if (lbClientCache.containsKey(clientName)) {
```

Comment:
actually, double-check locking won't save us. ex. cache.get could read a corrupt table if another client is being inserted (ex. during the lookup of client1, client2 is being inserted). put the whole shebang in a lock.

```java
synchronized(cache) {
  LBClient result = cache.get(clientName);
  if (result != null) return result;
  
  IClientConfig config = ClientFactory.getNamedConfig(clientName);
  ILoadBalancer lb = ClientFactory.getNamedLoadBalancer(clientName);
  result = new LBClient(lb, config);

  cache.put(clientName, result);
  return result;
}
```